### DISCUSSION TITLE
I'm beating my brains out trying to get Pandoc to work on Windows

### DISCUSSION THREAD

**Discussion by seatrout** [Taitava/obsidian-shellcommands#363] [Q&A] [answered]
Using powershell core, I am trying to write a shell command to reproduce this, which works in the terminal:
```
pandoc.exe C:\Users\Andrew\Vaults\inbox\QuickNote.md -o C:\Users\Andrew\Vaults\inbox\QuickNote.docx  
--resource-path=C:\Users\Andrew\AppData\Roaming\Pandoc 
--reference-doc=mycustom.docx
```
(reformatted for clarity, but all on one line in the original.)

Turned into a shell command, it becomes 

```
start 'C:\Program Files\Pandoc\pandoc.exe' "{{file_path:absolute}}
-o {{folder_path:absolute}}/{{title}}.docx
--resource-path=C:\Users\Andrew\AppData\Roaming\Pandoc
--reference-doc=mycustom.docx
{{file_path:absolute}}"
```

And when I run this, nothing happens — at least, nothing is written to disk by pandoc. I have tried every possible combination of quotes, single and double, of running it with and without "start" at the beginning, passing the arguments through stdin ... 

Has anyone any working examples of pandoc running under powershell in windows?

**Comment by seatrout** [comment]
Ah. The absolute simplest version *does* work:
`start 'C:\Program Files\Pandoc\pandoc.exe' "{{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.docx"`
But all the parameters I want to pass in don't work, and result in a silent failure to write to disk. I will have a look at the pandoc documentation

[Edit] No, no it doesn't work (again) when the file to be converted has spaces in the name. Again, a silent failure. Nothing is written to disk.

**Reply by seatrout** [reply]
It turns out that easiest(!?) way to do this is to drop the whole problem into wsl. Everything there works the way that it ought to do from the documentation. But there are obviously huge problems in passing parameters to commands in powershell when these commands will only work using "start" as a prefix.

**Comment by seatrout** [comment, marked as answer]
OK. In the hope this will be useful to others, here is the answer I have found that works. You need to **de-escape** the file and folder names, and pass them as **literal strings,** like this:

`start pandoc.exe '"{{!file_path:absolute}}" -o "{{!folder_path:absolute}}\{{!title}}.docx"'`

so: _Single Quotes_ round the whole string of parameters but within those single quotes, _double quotes_ round the _unescaped_ parameter values (not the keys). The command itself remains unquoted.

**Reply by Taitava** [reply]
Hi! 🙂 Good that you seem to have found a solution. 👍 

In case you still want to try to simplify your command, you can try these two suggestions:

```
pandoc.exe "{{!file_path:absolute}}" -o "{{!folder_path:absolute}}\{{!title}}.docx"
```

Or:


```
pandoc.exe {{file_path:absolute}} -o {{folder_path:absolute}}\{{!title}}.docx
```

The latter does not need any quotes, because possible spaces are automatically escaped by preceding them with backticks ` 

Does this bring any benefit for you? 🙂

**Reply by seatrout** [reply]
I did try both of those when I was beating my brains out. The trouble is the `start (Start-Process) command` which is needed in Powershell for Windows. This adds a whole layer of indirection and complication to the parameters passed to pandoc. That is why the whole string of pandoc parameters has to be enclosed in single quotes. Without those quotes, it is `start` which processes the parameters for pandoc, and it barfs on the `-o`. 

I do wish Windows had the same escape conventions as linux but there we are.

**Reply by seatrout** [reply]
To be clear — your suggestions **_ought_** to work, but Powershell does not seem to like or understand escaped Windows paths reliably. So I have to pass them in unescaped, within double quotes, if I want them to work.

### An example: 
I write my markdown files with one carriage return at the end of every paragraph, not two. I know this isn't strictly the Markdown spec, but Obsidian understands it, and it's how I have typed for thirty years now. Think of the keystrokes I've saved! 
Pandoc wants two line ends between paragraphs, though. 
So if I want to send a markdown file to word, the first thing to do is to double all the paragraph breaks and then send it to pandoc.
#### The shell command to do this in WSL (or any other *nix) is nice and simple:

```
sed 's/$/\n/' < {{file_path:absolute}} > {{folder_path:absolute}}\/temp\.md &&
pandoc {{folder_path:absolute}}\/temp.md -o {{folder_path:absolute}}\/{{title}}\.docx &&
rm {{folder_path:absolute}}\/temp\.md
```

There are single quotes round the `sed` parameters and that's it. 
#### In Windows, though ... 

```
'$singlespace = get-content "{{!file_path:absolute}}" -Raw'
'$doublespace = $singlespace -replace "`n","`n`n"'
'Set-Content -Path doublespaced.md -Value $doublespace'
start pandoc 'doublespaced.md -o "{{!folder_path:absolute}}\{{!title}}.docx"'
'Remove-Item doublespaced.md'
```

Apart from the verbosity of the commands, there needs to be single quotes round almost everything — all the commands and the parameters to pandoc. I don't know why. But I do know that this works, and this is the example I will put in my fork of the documentation.

**Reply by Taitava** [reply]
> Without those quotes, it is `start` which processes the parameters for pandoc, and it barfs on the `-o`.

Ah, I see. But I'm just wondering, why is `start` even needed on Windows?

**Reply by Taitava** [reply]
> To be clear — your suggestions **_ought_** to work, but Powershell does not seem to like or understand escaped Windows paths reliably.

Does this test work in PowerShell:
```
type {{file_path:absolute}}
```
It should output the currently open file's content to `stdout`. If you hit Ctrl + _Execute now_ icon in SC settings, you should see the output (i.e. the file's content) in a modal. If it can output the file's content, then PowerShell understands escaped paths at least in this kind of simple shell command.

> ```
> '$singlespace = get-content "{{!file_path:absolute}}" -Raw'
> ```

I tried to execute that single line on my environment (Obsidian + SC + PowerShell) and it didn't assign content to the variable, it just echoed the statement back, but without the wrapping single quotes:
```
$singlespace = get-content "V:\Lajitellut\Obsidian-tietokannat\Shell commands test\Sandbox\TestResults.md" -Raw
```

I wonder how it works in your environment? When you use PowerShell Core, do you use the one that is built-in to the _Shell commands_ plugin, or have you defined a [custom shell](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Custom+shells) for PowerShell?

**Reply by seatrout** [reply]
Your test works perfectly well in Powershell Core, which is what I am using.

**Reply by seatrout** [reply]
"Start" is needed with the pandoc command because `pandoc` doesn't work without it :-)

**Reply by Taitava** [reply]
But is there a specific reason, why `pandoc` does not work without `start`? 🙂

**Comment by seatrout** [comment]
Ah!!! I am using Powershell 7.3 

It never occurred to me that there might be a powershell version built into SC. 

There must be some difference in quoting behaviour between the two. This is going to need a little thought. I can't be the only person to have been misled.

**Reply by Taitava** [reply]
If you want, you can post your SC settings file (`.obsidian/plugins/obsidian-shellcommands/data.json`) here, so I can have a better look on how you've configured PowerShell 🙂 . You can remove any possible private information from it first, if there's something you don't want to share publicly.

**Comment by seatrout** [comment]
[data.json zipped for Taitava.zip](https://github.com/Taitava/obsidian-shellcommands/files/12431446/data.json.zipped.for.Taitava.zip)

No secrets here! and you can see a couple of experimental commands in there just to check other things are working — which they are.

**Reply by Taitava** [reply]
I tried your settings file now on my Windows 10 machine, with PowerShell 7.3.6.

For me, this didn't work:
```PowerShell
'$singlespace = get-content "{{!file_path:absolute}}" -Raw'
'$doublespace = $singlespace -replace "`n","`n`n"'
'Set-Content -Path doublespaced.md -Value $doublespace'
start pandoc 'doublespaced.md -o "{{!folder_path:absolute}}\{{!title}}.docx"'
'Remove-Item doublespaced.md'
```

It just outputted the text that was in single quotes:
```
$singlespace = get-content "V:\Lajitellut\Obsidian-tietokannat\Shell commands test\Sandbox\TestResults.md" -Raw
$doublespace = $singlespace -replace "`n","`n`n"
Set-Content -Path doublespaced.md -Value $doublespace
Remove-Item doublespaced.md
```

I changed the command to this, and it works perfectly fine for me:
```PowerShell
$singlespace = get-content "{{!file_path:absolute}}" -Raw
$doublespace = $singlespace -replace "`n","`n`n"
Set-Content -Path doublespaced.md -Value $doublespace
pandoc doublespaced.md -o "{{!folder_path:absolute}}\{{!title}}.docx"
Remove-Item doublespaced.md
```

> It never occurred to me that there might be a powershell version built into SC.

What I actually meant about a "builtin shell" was a ready-made configuration, so that users don't have to define a custom shell when they want to use PowerShell. I might have been confusing with my words 🙂. Anyway, you seem to be using the "builtin" PowerShell Core. 👍 

---

So, for a conclusion: I wasn't able to find out why the single quotes are needed around PowerShell commands in your case, nor why is `start` needed. For me, `pandoc` worked and was able to produce a docx file. So, unfortunately I can't seem to be able to solve the issues you are having with PowerShell. I hope you can deal with doing stuff mostly on WSL, if it works better for you than PowerShell. 🙂
